src/core/algorithms/nar/value_range.h

```diff
@@ -0,0 +1,94 @@
+#pragma once
+
+#include <enum.h>
+
+#include "model/table/column_layout_typed_relation_data.h"
+#include "model/types/type.h"
+
+namespace model {
+
+class ValueRange {
+protected:
+    ValueRange() {};
+    virtual ~ValueRange() = default;
+
+public:
+    virtual TypeId GetTypeId() const = 0;
+    virtual bool Includes(std::byte const* value) const = 0;
+    virtual std::string ToString() const = 0;
+};
+
+class StringValueRange : public ValueRange {
+public:
+    std::vector<String> domain{};
+
+    explicit StringValueRange(TypedColumnData const& column);
+    StringValueRange(String value) : domain{std::move(value)} {};
+    StringValueRange(std::vector<String> vec) : domain{std::move(vec)} {};
+
+    TypeId GetTypeId() const override {
+        return TypeId::kString;
+    }
+
+    bool Includes(std::byte const* value) const override {
+        String const& svalue = Type::GetValue<String>(value);
+        return std::find(domain.begin(), domain.end(), svalue) != domain.end();
+    }
+
+    std::string ToString() const override;
+    ~StringValueRange() = default;
+};
+
+template <typename T>
+class NumericValueRange : public ValueRange {
+public:
+    T lower_bound{};
+    T upper_bound{};
+
+    explicit NumericValueRange(TypedColumnData const& column) {
+        bool initialized = false;
+        for (size_t row_index = 0; row_index < column.GetNumRows(); ++row_index) {
+            std::byte const* value = column.GetValue(row_index);
+            T numeric_value = Type::GetValue<T>(value);
+            if (!initialized) {
+                lower_bound = numeric_value;
+                upper_bound = numeric_value;
+                initialized = true;
+            } else {
+                if (numeric_value < lower_bound) {
+                    lower_bound = numeric_value;
+                }
+                if (numeric_value > upper_bound) {
+                    upper_bound = numeric_value;
+                }
+            }
+        }
```

Comment:
```suggestion
        std::vector<T> values(column.GetNumRows());

        for (std::byte const* raw_value : column.GetData()) {
            values.push_back(Type::GetValue<T>(raw_value));
        }

        auto [min, max] = std::ranges::minmax(values);
        lower_bound = min;
        upper_bound = max;
```
Smth like this?

Reply:
sorry, I meant reserve here instead of a constructor from int 
```cpp
std::vector<T> values();
values.reserve(column.GetNumRows());
```

Reply:
Duplicating the whole column into the `values` vector seems wasteful to me. What if we have a dataset that is 5 columns, each with 5000 rows. Then we are duplicating 20% of the dataset, when the rest of the algo can be made non-memory-demanding if population is set to some low value.

Reply:
yeah, makes sense, thanks

Reply:
hm, I think this will work then, cause we can just tell `std::minmax` how to extract values:
```cpp
    auto [min_ptr, max_ptr] = std::ranges::minmax(column.GetData(), {},
        [](std::byte const* value) { 
            return Type::GetValue<T>(value); 
        });

    lower_bound = Type::GetValue<T>(min_ptr);
    upper_bound = Type::GetValue<T>(max_ptr);
```

Reply:
works but makes the test run 14 seconds instead of 1 :)

Reply:
![image](https://github.com/user-attachments/assets/7bea6ad6-7f17-4767-823a-b12ae0b31f68)

Reply:
That's unfortunate :skull: 
Because all that is just one `std::minmax` algo, don't want to give up on that and use raw loop.. But, well, I see no other choice.

But still let's use `std::min` and `std::max` then instead of comparing manually.

Like this?
```cpp
    std::byte const* first_value = column.GetValue(0);
    T lower_bound = Type::GetValue<T>(first_value);
    T upper_bound = lower_bound;

    for (size_t row_index = 1; row_index < column.GetNumRows(); ++row_index) {
        std::byte const* value = column.GetValue(row_index);
        T numeric_value = Type::GetValue<T>(value);

        lower_bound = std::min(lower_bound, numeric_value);
        upper_bound = std::max(upper_bound, numeric_value);
    }
```
(don't want to add extra branch in a loop)

Reply:
If we really wanted to make it more concise we could do something like this, getting rid of the `initialized` bool:

```C++
template <typename T>
struct NumericValueRange : public ValueRange {
    T lower_bound = std::numeric_limits<T>::max();
    T upper_bound = std::numeric_limits<T>::lowest();

    explicit NumericValueRange(TypedColumnData const& column) {
        for (size_t row_index = 0; row_index < column.GetNumRows(); ++row_index) {
            std::byte const* value = column.GetValue(row_index);
            T numeric_value = Type::GetValue<T>(value);
            if (numeric_value < lower_bound) {
                lower_bound = numeric_value;
            }
            if (numeric_value > upper_bound) {
                upper_bound = numeric_value;
            }
        }
    }
```

Reply:
edited^